File: spark/src/main/scala/com/intel/genomicsdb/GenomicsDBRecordReader.scala
package com.intel.genomicsdb

import genomicsdb.GenomicsDBFeatureReader
import htsjdk.tribble.{CloseableTribbleIterator, Feature}
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.mapreduce.{InputSplit, RecordReader, TaskAttemptContext}

import scala.reflect.ClassTag

/**
  * A RecordReader class to read variant records from GenomicsDB.
  * It emits serializable vatiant contexts so that RDDs can use
  * them. The type of keys are NULL, and the type of values are
  * VariantContexts
  *
  * @param split
  * @param conf
  * @tparam D
  * @tparam S
  */
class GenomicsDBRecordReader[D <: Feature: ClassTag, S: ClassTag](
    split: GenomicsDBInputSplit,
    conf: GenomicsDBConf,
    featureReader: GenomicsDBFeatureReader[D, S]) extends RecordReader[LongWritable, D] {

  var currentVariant : D = _
  var iterator: CloseableTribbleIterator[D] = _

  @Override
  def getProgress: Float = {
    0
  }

  //def nextKeyValue: D = iterator.next()

  @Override
  def getCurrentValue: D = {
    currentVariant
  }

  @Override
  def initialize(inputSplit: InputSplit, taskAttemptContext: TaskAttemptContext): Unit = {
    initialize
  }

  def initialize: Unit = {
    iterator = featureReader.iterator
    if (hasNext)
      currentVariant = iterator.next
  }

  @Override
  def getCurrentKey: LongWritable = new LongWritable(currentVariant.toString.hashCode.toLong)

  @Override
  def close(): Unit = {
    iterator.close()
  }

  override def nextKeyValue(): Boolean = {
    if (hasNext) true else false
  }

  def nextValue: D = iterator.next

  def hasNext: Boolean = iterator.hasNext
}

File: spark/src/main/java/com/intel/genomicsdb/GenomicsDBRecordReader.java
package com.intel.genomicsdb;

import genomicsdb.GenomicsDBFeatureReader;
import htsjdk.tribble.CloseableTribbleIterator;
import htsjdk.tribble.Feature;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.mapreduce.InputSplit;
import org.apache.hadoop.mapreduce.RecordReader;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.log4j.Logger;

import java.io.IOException;

public class GenomicsDBRecordReader<VCONTEXT extends Feature, SOURCE>
  extends RecordReader<LongWritable, VCONTEXT> {

  private final GenomicsDBFeatureReader<VCONTEXT, SOURCE> featureReader;
  private CloseableTribbleIterator<VCONTEXT> iterator;
  private VCONTEXT currentVariant;
  private int currentKey;

  Logger logger = Logger.getLogger(GenomicsDBRecordReader.class);

  GenomicsDBRecordReader(GenomicsDBFeatureReader<VCONTEXT, SOURCE> featureReader) {
    this.featureReader = featureReader;
    this.currentKey = -1;
    logger = Logger.getLogger(GenomicsDBRecordReader.class);
  }

  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
    throws IOException, InterruptedException {
    initialize();
  }

  private void initialize() throws IOException {
    this.iterator = featureReader.iterator();
    if (iterator.hasNext()) {
      this.currentVariant = iterator.next();
      this.currentKey++;
    }
  }

  public boolean nextKeyValue() throws IOException, InterruptedException {
    if (this.iterator.hasNext()) {
      this.currentVariant = iterator.next();
      this.currentKey++;
      return true;
    } else return false;
  }

  public LongWritable getCurrentKey() throws IOException, InterruptedException {
    return new LongWritable(currentKey);
  }

  public VCONTEXT getCurrentValue() throws IOException, InterruptedException {
    return currentVariant;
  }

  public float getProgress() throws IOException, InterruptedException {
    return 0;
  }

  public void close() throws IOException {
    this.iterator.close();
  }

  public Boolean hasNext() {
    return this.iterator.hasNext();
  }

  public VCONTEXT nextValue() {
    return iterator.next();
  }
}
